File: Java Labs/Patterns On Java/src/com/pattern/memento/ShoppingCart.java
package com.pattern.memento;

import java.util.ArrayList;
import java.util.List;

public class ShoppingCart {
    private List<String> items;

    public ShoppingCart() {
        this.items = new ArrayList<>();
    }

    public void addItem(String item) {
        items.add(item);
    }

    public void displayItems() {
        System.out.println("Shopping Cart Items: " + items);
    }

    public ShoppingCartMemento save() {
        return new ShoppingCartMemento(new ShoppingCartState(items));
    }

    public void restore(ShoppingCartMemento memento) {
        this.items = memento.getState().getItems();
    }
}

File: Python Labs/com/pattern/memento/ShoppingCart.py
from com.pattern.memento.ShoppingCartMemento import ShoppingCartMemento
from com.pattern.memento.ShoppingCartState import ShoppingCartState


class ShoppingCart:
    def __init__(self):
        self.items = []

    def add_item(self, item):
        self.items.append(item)

    def display_items(self):
        print("Shopping Cart Items:", self.items)

    def save(self):
        return ShoppingCartMemento(ShoppingCartState(self.items))

    def restore(self, memento):
        self.items = memento.get_state().get_items()